### ISSUE TITLE
How does StormLoader actually load mods?

### ISSUE THREAD

**Issue by AnonymousDapper** [Lewinator56/StormLoader#18]
In my experiments I've been unable to get Stormworks to want to load mods when I manually copy them into the game directory.
However, compressing the raw files and installing the zip with StormLoader works flawlessly.

What method do you use to convince the game to load external resources? I've poked around the codebase a bit but I have not found anything that appears to be doing it.

**Comment by Lewinator56** [comment]
Thanks for your interest, you can see most of the install code here [github.com/lewinator56/StormLoader/StormLoader/mod_handling/ModInstaller.cs ] (https://github.com/Lewinator56/StormLoader/blob/master/StormLoader/StormLoader/mod_handling/ModInstaller.cs)

It is simply copying data into the game directory, however because stormworks has no official support for modding, each mod has to be broken up and it's files distributed around to the appropriate directories in the game folder. This of course creates an issue for tracking installs - I get around this by storing details about all the non-vanilla files installed in a small binary file in the stormworks directory. There's also a number of limitations imposed by the game, for example, it only looks at the top level directory for part definitions, but can look in subdirectories for meshes, which can lear to quite a confusing file layout for some mods.

**Comment by AnonymousDapper** [comment]
So StormLoader doesn't do anything besides just copy the mod files (meshes, audio, definitions, etc)?

When manually copying files into the game directory it's as though I haven't even changed anything; the game refuses to acknowledge the new files.

However, just by repacking and installing via StormLoader the mods work fine. Any idea what might be causing this?

(And how I can get mods to load on my Mac where I don't have StormLoader?)

**Comment by Lewinator56** [comment]
Correct, for all intents and purposes it's a glorified copy and paste program - obviously the tracking and load order stuff on top, I've been debating about trying to introduce a virtual file system through the use of symbolic links, a bit like what nexus mod manager does.

As for installing mods on a Mac, I know that stormloader can't run on one, (and never will), however provided the files are placed in the right place I can't see why there should be any issue with the game loading them. You should simply be able to locate the game package and browse it's files, you can then just copy the stuff in to the right places.

For manual installs on windows you may be having issues if you are putting definitions in a subfolder in the definitions root folder - as I previously stated, stormworks only looks in the topmost folder for part definitions. A mod is not simply just a mesh, it requires an accompanying definition XML to tell the game about collision bounds, logic nodes, buoyancy etc... So a part consists of both a .mesh and a .xml which both have to be put in the right places in the games directory. It sounds like you've already got the part where you need 2 files sorted, so it sounds to me like you are copying something into the wrong folder.

**Comment by Lewinator56** [comment]
Obviously, simplifying the inner workings of stormloader down to simply a copy and paste tool kind of undermines it. It was built as an accessibility tool for users who don't care about putting files in the right places, but simply want something to manage it all for them - a key part of enabling modding for stormworks was making it accessible for the majority of the playerbase (who are, let's face it, relatively young) - mainly to disprove the developers' concerns about modding not being accessible. With this in mind, the fact a user can open stormloader, browse and online repo and click 1 button to download an install a mod, I think accessibility is good. The recent updates introduce support for mods uploaded to the steam workshop, using an approach that creates a custom mission containing mod content which stormloader can read at startup. Along with introducing a built in packager that can generate these steam workshop compatible mods to help make the creation of mods easier to understand (people have trouble with the SLP format sometimes).

Of course there are a number of things I want to add - one of which is automatic steam mod installation at game launch. This would require replacing the stormworks executable with one that first checked the special missions packed for stormloader, call stormloader to install them, then launch the game - as far as a user is concerned this all happens in the background and would be the closest thing to official mods that I as a developer could do.
